pkg/api/dirs_test.go

```diff
@@ -0,0 +1,247 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package api_test
+
+import (
+	"archive/tar"
+	"bytes"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path"
+	"testing"
+
+	"github.com/ethersphere/bee/pkg/api"
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/jsonhttp/jsonhttptest"
+	"github.com/ethersphere/bee/pkg/logging"
+	"github.com/ethersphere/bee/pkg/manifest/jsonmanifest"
+	"github.com/ethersphere/bee/pkg/storage/mock"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/tags"
+)
+
+type dirTestCase struct {
+	expectedHash string
+	files        []dirTestCaseFile
+}
+
+type dirTestCaseFile struct {
+	data      []byte
+	name      string
+	path      string
+	reference swarm.Address
+	headers   http.Header
+}
+
+func TestDirs(t *testing.T) {
+	var (
+		dirUploadResource    = "/dirs"
+		fileDownloadResource = func(addr string) string { return "/files/" + addr }
+		client               = newTestServer(t, testServerOptions{
+			Storer: mock.NewStorer(),
+			Tags:   tags.NewTags(),
+			Logger: logging.New(ioutil.Discard, 5),
+		})
+	)
+
+	t.Run("empty request body", func(t *testing.T) {
+		jsonhttptest.ResponseDirectSendHeadersAndReceiveHeaders(t, client, http.MethodPost, dirUploadResource, bytes.NewReader(nil), http.StatusBadRequest, jsonhttp.StatusResponse{
+			Message: "could not extract dir info from request",
+			Code:    http.StatusBadRequest,
+		}, nil)
+	})
+
+	t.Run("empty file", func(t *testing.T) {
+		f := writeAndOpenFile(t, "empty-file", nil)
+		defer os.Remove(f.Name())
+		defer f.Close()
+
+		jsonhttptest.ResponseDirectSendHeadersAndReceiveHeaders(t, client, http.MethodPost, dirUploadResource, f, http.StatusInternalServerError, jsonhttp.StatusResponse{
+			Message: "could not store dir",
+			Code:    http.StatusInternalServerError,
+		}, nil)
+	})
+
+	t.Run("non tar file", func(t *testing.T) {
+		f := writeAndOpenFile(t, "non-tar-file", []byte("some data"))
+		defer os.Remove(f.Name())
+		defer f.Close()
+
+		jsonhttptest.ResponseDirectSendHeadersAndReceiveHeaders(t, client, http.MethodPost, dirUploadResource, f, http.StatusInternalServerError, jsonhttp.StatusResponse{
+			Message: "could not store dir",
+			Code:    http.StatusInternalServerError,
+		}, nil)
+	})
+
+	for _, tc := range []dirTestCase{
```

Comment:
inlining the test case data struct into the for loop makes things more readable, but also writeable, since you dont need to scroll all the time to look for the struct definition:
```go
for _,tc := range []struct{
 name string
 field int...
} {
 { name: "test1", field: 123},
 { name: "test2", field: 456}, ...
} {
 //do stuff
}
```

Reply:
thanks. i've changed this, but it looks a little weird by having to define—what previously was—the `dirTestCaseFile` type twice, since it is inline now.

is there a way around this? am i missing something?

i've looked around and did not find anything explicit for this.

Reply:
i did not mean `dirTestCaseFile`, i meant `dirTestCase`. for `dirTestCaseFile` you can try to employ a similar approach to what i did in for example `pullsync_test.go` with `someChunks` or in `puller_test.go` with `expLiveCalls`. can't think of anything else but it surely increases readability

Reply:
thanks. went with something like what was done in `puller_test.go`.

it's definitely more readable now